src/slist.c

```diff
@@ -1 +1,297 @@
+
+		while (list_instance->next != NULL)
+		{
+			list_instance = list_instance->next;
+		}
+
+		list_instance->next = list_entry;
+		result = list_entry;
+	}
+	return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_add_head(PSINGLYLINKEDLIST_ENTRY list_head, PSINGLYLINKEDLIST_ENTRY list_entry)
+{
+	PSINGLYLINKEDLIST_ENTRY result;
+	if ((list_head == NULL) ||
+		(list_entry == NULL))
+	{
+		/* Codes_SRS_SLIST_07_008: [ If `list_head` is `NULL`, `slist_add_head` shall fail and return `NULL`. ]*/
+		/* Codes_SRS_SLIST_07_009: [ If `list_entry` is `NULL`, `slist_add_head` shall fail and return `NULL`. ]*/
+		LogError("Invalid argument (list_head=%p, list_entry=%p)", list_head, list_entry);
+		result = NULL;
+	}
+	else
+	{
+		/* Codes_SRS_SLIST_07_010: [ `slist_add_head` shall insert `list_entry` at head on success and return an entry to the added entry. ]*/
+		list_entry->next = list_head;
+		result = list_entry;
+	}
+	return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_remove(PSINGLYLINKEDLIST_ENTRY list_head, PSINGLYLINKEDLIST_ENTRY list_entry)
+{
+	PSINGLYLINKEDLIST_ENTRY result;
+	if ((list_head == NULL) ||
+		(list_entry == NULL))
+	{
+		/* Codes_SRS_SLIST_07_011: [ If `list_head` is `NULL`, `slist_remove` shall fail and return `NULL`. ]*/
+		/* Codes_SRS_SLIST_07_012: [ If `list_entry` is `NULL`, `slist_remove` shall fail and return `NULL`. ]*/
+		LogError("Invalid argument (list_head=%p, list_entry=%p)", list_head, list_entry);
+		result = NULL;
+	}
+	else
+	{
+		PSINGLYLINKEDLIST_ENTRY prev_item = NULL;
+		PSINGLYLINKEDLIST_ENTRY current_item = list_head;
+		while (current_item != NULL)
+		{
+			if (current_item == list_entry)
+			{
+				break;
+			}
+			prev_item = current_item;
+			current_item = current_item->next;
+		}
+		if (current_item == NULL)
+		{
+			/* Codes_SRS_SLIST_07_014: [ If the entry `list_entry` is not found in the list, then `slist_remove` shall fail and `NULL`. ]*/
+			result = NULL;
+		}
+		else 
+		{
+			/* Codes_SRS_SLIST_07_013: [ `slist_remove` shall remove a list entry from the list on success and return a pointer to the new head entry. ]*/
+			if (prev_item != NULL)
+			{
+				result = list_head;
+				prev_item->next = current_item->next;
+			}
+			else
+			{
+				result = current_item->next;
+			}
+		}
+	}
+	return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_remove_head(PSINGLYLINKEDLIST_ENTRY list_head)
+{
+	PSINGLYLINKEDLIST_ENTRY result;
+	if (list_head == NULL)
+	{
+		/* Codes_SRS_SLIST_07_015: [ If `list_head` is `NULL`, `slist_remove_head` shall fail and return `NULL`. ]*/
+		LogError("Invalid argument (list_head=%p)", list_head);
+		result = NULL;
+	}
+	else
+	{
+		/* Codes_SRS_SLIST_07_016: [ `slist_remove_head` removes the head entry from the list defined by the `list_head` parameter on success and return a pointer to the new head entry. ]*/
+		/* Codes_SRS_SLIST_07_017: [ `slist_remove_head` shall return `NULL` if that's the only entry in the list. ]*/
+		result = list_head->next;
```

Comment:
Whose responsibility is to free the memory of the removed entry?

Reply:
caller